src/linker/Linker.Dataflow/MethodBodyScanner.cs

```diff
@@ -779,16 +799,45 @@ void ScanLdtoken (Instruction operation, Stack<StackSlot> currentStack)
 		private void ScanIndirectStore (
 			Instruction operation,
 			Stack<StackSlot> currentStack,
-			MethodBody methodBody)
+			MethodBody methodBody,
+			LocalVariableStore locals,
+			int curBasicBlock)
 		{
 			StackSlot valueToStore = PopUnknown (currentStack, 1, methodBody, operation.Offset);
 			StackSlot destination = PopUnknown (currentStack, 1, methodBody, operation.Offset);
 
-			foreach (var uniqueDestination in destination.Value) {
-				if (uniqueDestination is FieldValue fieldDestination) {
-					HandleStoreField (methodBody.Method, fieldDestination, operation, valueToStore.Value);
-				} else if (uniqueDestination is MethodParameterValue parameterDestination) {
-					HandleStoreParameter (methodBody.Method, parameterDestination, operation, valueToStore.Value);
+			StoreReference (destination.Value, valueToStore.Value, methodBody.Method, operation, locals, curBasicBlock);
+		}
+
+		protected void StoreReference (MultiValue target, MultiValue source, MethodDefinition method, Instruction operation, LocalVariableStore locals, int curBasicBlock)
+		{
+			foreach (var value in target) {
+				switch (value) {
+				case LocalVariableReferenceValue localReference:
+					StoreMethodLocalValue (locals, source, localReference.LocalDefinition, curBasicBlock);
+					break;
+				case FieldReferenceValue fieldReference
+				when GetFieldValue (fieldReference.FieldDefinition).AsSingleValue () is FieldValue fieldValue:
+					HandleStoreField (method, fieldValue, operation, source);
+					break;
+				case ParameterReferenceValue parameterReference
+				when GetMethodParameterValue (parameterReference.MethodDefinition, parameterReference.ParameterIndex) is MethodParameterValue parameterValue:
+					HandleStoreParameter (method, parameterValue, operation, source);
+					break;
+				case ParameterReferenceValue parameterReference
+					when GetMethodParameterValue (parameterReference.MethodDefinition, parameterReference.ParameterIndex) is MethodThisParameterValue thisParameterValue:
+					HandleStoreMethodThisParameter (method, thisParameterValue, operation, source);
+					break;
+				case MethodReturnValue methodReturnValue:
+					HandleStoreMethodReturnValue (method, methodReturnValue, operation, source);
+					break;
+				case ArrayElementReferenceValue:
+				// Array elements are set to Unknown when a reference is taken, we don't track dataflow through array element references yet
+				case UnknownValue:
+					// If we don't know where the value is being stored, should we warn?
```

Comment:
Try to find a repro where this is hit... it's hard to reason about it if we don't know what the code would look like.

In theory this could be problematic if there was a way to assign a Type value to a place which has annotations on it. So we should probably warn, but we should try to figure out what a repro would look like.

Reply:
Since we return an unknown value for `ldelema`, we hit the unknown case when we pass a ref to an element as a ref parameter. I need to remove the ArrayElementReferenceValue case.

Reply:
If it's just arrays than that's fine - we don't support annotating `Type[]` so no problem there (specifically if you read a value from such location and it comes out annotated that would be a problem).